### PR TITLE
Allow `ThrottledFunc` to work on different types of dispatcher

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

This updates our `ThrottledFunc`s to take a dispatcher parameter. This means that we can use the `Windows::UI::Core::CoreDispatcher` in the `TermControl`, where there's always a `CoreDispatcher`, and use a `Windows::System::DispatcherQueue` in `ControlCore`/`ControlInteractivity`. When running in-proc, these are always the _same thing_. However, out-of-proc, the core needs a dispatcher queue that's not tied to a UI thread (because the content proces _doesn't have a UI thread!_). 

This lets us get rid of the output event, because we don't need to bubble that event out to the `TermControl` to let it throttle that update anymore. 

## References
* Tear-out: #1256
* Megathread: #5000
* Project: https://github.com/microsoft/terminal/projects/5

## PR Checklist
* [x] This is a part of #1256
* [x] I work here
* [n/a] Tests added/passed
* [n/a] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

Fortunately, `winrt::resume_foreground` works the same on both a `CoreDispatcher` and a `DispatcherQueue`, so this wasn't too hard!

## Validation Steps Performed

This was validated in `dev/migrie/oop/the-whole-thing` (or `dev/migrie/oop/connection-factory`, I forget which), and I made sure that it worked both in-proc and x-proc. Not only that, _it wasn't any slower_!This reverts commit 04b751faa70680bf0296063deacec4657c6ff9d6.

### CODE REVIEW

**Comment by DHowett** [comment]
Here's a crazy thought: What if it _only_ worked with the system dispatcher queue? It should be trivial to move TermControl to that.

**Comment by DHowett** [comment]
I don't love that this PR includes _moving the actual callers to a lower layer_, but i can learn to live with it.

**Comment by zadjii-msft** [comment]
> Here's a crazy thought: What if it _only_ worked with the system dispatcher queue? It should be trivial to move TermControl to that.

this is a _good idea_

**Comment by miniksa** [comment]
> > Here's a crazy thought: What if it _only_ worked with the system dispatcher queue? It should be trivial to move TermControl to that.
> 
> this is a _good idea_

I was going to review... but @zadjii-msft does your emphasis of "good idea" mean you're preparing an iteration that does this? Or is it future work?

**Comment by DHowett** [comment]
He confirms: yes, he was.

**Comment by zadjii-msft** [comment]
I don't _think_ I broke anything in this merge

**Comment by DHowett** [comment]
The diff look a little noisy, like it's bringing some stuff from `main` into `10051`

**Comment by zadjii-msft** [comment]
what the heck happened here, I swear I pushed - I'll fix it after lunch

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 1]
```diff
@@ -94,6 +104,61 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+        // Get our dispatcher. If we're hosted in-proc with XAML, this will get
+        // us the same dispatcher as TermControl::Dispatcher(). If we're out of
+        // proc, this'll return null. We'll need to instead make a new
+        // DispatcherQueue (on a new thread), so we can use that for throttled
+        // functions.
+        _dispatcher = winrt::Windows::System::DispatcherQueue::GetForCurrentThread();
+        if (!_dispatcher)
+        {
+            auto controller{ winrt::Windows::System::DispatcherQueueController::CreateOnDedicatedThread() };
+            _dispatcher = controller.DispatcherQueue();
```
Comment (miniksa) [inline src/cascadia/TerminalControl/ControlCore.cpp:116]: This is mildly confusing. The `controller` object goes out of scope and presumably a destructor is called... but the controller and thread stay alive anyway (presumably) as the child `DispatcherQueue()` is still valid... yeah?

Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.cpp:116]: Apparently, yea

**Review by miniksa — APPROVED** [review]

File: src/cascadia/TerminalControl/ControlCore.h [hunk 2]
```diff
@@ -239,6 +244,11 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         void _connectionOutputHandler(const hstring& hstr);
         void _updateHoveredCell(const std::optional<til::point> terminalPosition);
 
+        inline bool _IsClosing() const noexcept
+        {
+            return _closing.load();
+        }
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.h:250]: You can copy the variant from `TermControl` as is, allowing you to safely remove this incorrect use of an atomic.

Comment (lhecker) [reply src/cascadia/TerminalControl/ControlCore.h:250]: A suggestion:
```cpp
bool _closing{ false };

// ...

inline bool _IsClosing() const noexcept
{
#ifndef NDEBUG
	if (_dispatcher)
	{
	      // _closing isn't atomic and may only be accessed from the main thread.
	      assert(_dispatcher.HasThreadAccess());
	}
#endif
    return _closing;
}
```

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 3]
```diff
@@ -1078,15 +1143,21 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+        // Additionally, start the throttled update of where our links are.
+        _updatePatternLocations->Run();
     }
 
     void ControlCore::_terminalCursorPositionChanged()
     {
-        _CursorPositionChangedHandlers(*this, nullptr);
+        // When the buffer's cursor moves, start the throttled func to
+        // eventually dispatch a CursorPositionChanged event.
+        _tsfTryRedrawCanvas->Run();
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:1160]: I don't understand this change... So if I understand it correctly, this starts a throttled func to throw an `CursorPositionChanged` even every 100ms, right? But `TermControl` also got a `_tsfTryRedrawCanvas` which only triggers the update every 100ms. Now it's only updating every 200-300ms (depending on the race between the two timers). Shouldn't we remove either of the two throttled funcs then?

File: src/cascadia/TerminalControl/ControlCore.h (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.h]: good thinking

**Review by lhecker — APPROVED** [review]

File: src/cascadia/TerminalControl/ControlCore.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.cpp:1160]: Okay so you have a point here. I'm gonna keep the throttling on the core side, not the control side. The Core will want to throttle how often it hops across the COM boundary to tell the UI to update. The UI doesn't care so much - it should just update whenever the Core tells it to.

**Comment by zadjii-msft** [comment]
note to self:

Unittests are now failing because `ControlCore::_terminalScrollPositionChanged` triggers a throttledfunc to fire the `ScrollPositionChanged` callback. In real life, this is good - the user won't notice the delay. In the tests, this is bad - we want the callback firing instantly so we can check the value matches what we'd expect. Might need to introduce a delay in the tests to account for this. Unless I can think of a more clever solution.

**Pull request merged**